src/stream.ts

```diff
@@ -144,40 +147,57 @@ export class WebRTCStream implements Stream {
 
   private async _sinkFn(src: Source<Uint8ArrayList | Uint8Array>): Promise<void> {
     await this.opened.promise;
-    if (closed || this.writeClosed) {
+    if (this.streamState.state == StreamStates.CLOSED || this.streamState.state == StreamStates.WRITE_CLOSED) {
       return;
     }
 
-    let self = this;
-    let closeWriteIterable = {
+    const self = this;
+    const closeWriteIterable = {
       async *[Symbol.asyncIterator]() {
         await self.closeWritePromise.promise;
         yield new Uint8Array(0);
       },
     };
 
     for await (const buf of merge(closeWriteIterable, src)) {
-      if (closed || this.writeClosed) {
-        break;
+      const state = self.streamState.state;
+      if (state == StreamStates.CLOSED || state == StreamStates.WRITE_CLOSED) {
+        return;
+      }
+      const msgbuf = pb.Message.toBinary({message: buf.subarray()});
+      const sendbuf = lengthPrefixed.encode.single(msgbuf)
+      this.channel.send(sendbuf.subarray())
+    }
+  }
+
+  processIncomingProtobuf(buffer: Uint8Array): pb.Message {
+    const m = pb.Message.fromBinary(buffer);
+    if (m.flag) {
+      const [currentState, nextState] = this.streamState.transition({direction: 'inbound', flag: m.flag!});
+      if (currentState != nextState) {
+        switch (nextState) {
+          case StreamStates.READ_CLOSED:
+            this._innersrc.end();
+            break;
+          case StreamStates.WRITE_CLOSED:
+            this.closeWritePromise.resolve();
+            break;
+          case StreamStates.CLOSED:
+            this.close();
+        }
       }
-      let res = buf.subarray();
-      let send_buf = pb.Message.toBinary({ message: buf.subarray() });
-      log.trace(`[stream:${this.id}][${this.stat.direction}] sending message: length: ${res.length} ${res}, encoded through pb as ${send_buf}`);
-      this.channel.send(send_buf);
     }
+    return m;
```

Comment:
I'm somewhat surprised this isn't `return m.message;`
The caller isn't planning on dealing with the flag again, is it?